pandas/core/internals/construction.py

```diff
@@ -422,10 +422,10 @@ def to_arrays(data, columns, coerce_float=False, dtype=None):
 
 def _list_to_arrays(data, columns, coerce_float=False, dtype=None):
     if len(data) > 0 and isinstance(data[0], tuple):
-        content = list(lib.to_object_array_tuples(data).T)
+        content = list(lib.to_object_array_tuples(list(data)).T)
```

Comment:
can you move this down even futher into to_object_array

Reply:
Just to clarify: `to_object_array_tuples` is used to convert list of tuples, and `to_object_array` is used to convert list of lists. How could these two case be unified to `to_object_array` ?

The tuple (or list) needs to be converted to list before send to `to_object_array_tuples`, otherwise we will see incorrect type error for `pd.DataFrame(((),()))`.

Reply:
what i mean is push the listifying down to these 2 routines, in fact ``to_object_array`` already has this, need to add similar in ``to_object_array_tuples``

Reply:
In fact, the segmentation fault happens in #25691 is caused by `to_object_array`, rather than `to_object_array_tuples`. The `input_rows = <list>rows` does nothing than an assignment:

```
  /* "pandas/_libs/lib.pyx":2272
 *         list row
 * 
 *     input_rows = <list>rows             # <<<<<<<<<<<<<<
 *     n = len(input_rows)
 * 
 */
  __pyx_t_1 = __pyx_v_rows;
  __Pyx_INCREF(__pyx_t_1);
  __pyx_v_input_rows = ((PyObject*)__pyx_t_1);
  __pyx_t_1 = 0;
```

The `rows` and `input_rows` are both `PyObject *`. The whole story is, when we invoke `to_object_array` with `([], [])`, we use list's `[]`(__Pyx_GetItemInt_List_Fast), it results a null pointer.  Then we `Py_INCREF` it, thus the segmentation fault.

I have revised the patch to remove the unused conversion in `to_object_array` and restrict its argument type to `list`. Another possible solution should be change the type of argument `rows` to `object` from `list` the both two methods, but I think lists is used more commonly to construct DataFrame than tuples. `__Pyx_GetItemInt_List_Fast` should be faster than `__Pyx_GetItemInt_Generic` and the conversion from tuple to list in python side is fairly ok.

<details>
  <summary>The backtrace of the segmentation fault</summary>

```
>>> lib.to_object_array(([], []))

Thread 1 "python" received signal SIGSEGV, Segmentation fault.
__Pyx_GetItemInt_List_Fast (boundscheck=1, wraparound=1, i=0, o=([], [])) at pandas/_libs/lib.c:55829
55829           Py_INCREF(r);
(gdb) bt
#0  __Pyx_GetItemInt_List_Fast (boundscheck=1, wraparound=1, i=0, o=([], [])) at pandas/_libs/lib.c:55829
#1  __pyx_pf_6pandas_5_libs_3lib_102to_object_array (__pyx_self=<optimized out>, __pyx_v_min_width=<optimized out>,
    __pyx_v_rows=([], [])) at pandas/_libs/lib.c:30958
#2  __pyx_pw_6pandas_5_libs_3lib_103to_object_array (__pyx_self=<optimized out>, __pyx_args=<optimized out>,
    __pyx_kwds=<optimized out>) at pandas/_libs/lib.c:30855
```

</details>

Reply:
hmm, IIRC this was put in here pretty specifically, can you see why that was? in any event, I believe you can simply type the input arg to ``to_object_array_tuples`` no?

Reply:
In `to_object_array_tuples`, if the element in `rows` is not a tuple, it will be cast to a tuple.

https://github.com/pandas-dev/pandas/blob/master/pandas/_libs/lib.pyx#L2325-L2337

However I still think it make sense to treat `list of lists` and `list of tuples` differently and dispatch them to different methods to avoid the overhead caused by casting `list` to `tuple`. I think `list of lists` may be used more frequently. I have add a comment to `to_object_array_tuples`.

Reply:
@sighingnow I agree with your point, I am not suggesting combing these routines at all, rather *inside* each routine the data is first cast to a list

Reply:
Done.